cli/cmd/encore/deploy.go

```diff
@@ -0,0 +1,101 @@
+
+	"encr.dev/cli/cmd/encore/cmdutil"
+	"encr.dev/cli/internal/platform"
+	"encr.dev/pkg/appfile"
+)
+
+var (
+	appSlug string
+	envName string
+	commit  string
+	branch  string
+	format  string
+)
+
+var deployAppCmd = &cobra.Command{
+	Use:                   "deploy --commit COMMIT_SHA | --branch BRANCH_NAME",
+	Short:                 "Deploy an Encore app to a cloud environment",
+	DisableFlagsInUseLine: true,
+	Hidden:                true,
+	Run: func(c *cobra.Command, args []string) {
+		if commit != "" {
+			hb, err := hex.DecodeString(commit)
+			if err != nil || len(hb) != 20 {
+				cmdutil.Fatalf("invalid commit: %s", commit)
+			}
+		}
+		if appSlug == "" {
+			appRoot, _, err := cmdutil.MaybeAppRoot()
+			if err != nil {
+				cmdutil.Fatalf("no app found. Run deploy inside an encore app directory or specify the app with --app")
+			}
+			appSlug, err = appfile.Slug(appRoot)
+			if err != nil {
+				cmdutil.Fatalf("no app found. Run deploy inside an encore app directory or specify the app with --app")
+			}
+		}
+		if envName == "" {
+			envName = "@primary"
+		}
+		rollout, err := platform.Deploy(c.Context(), appSlug, envName, commit, branch)
+		var pErr platform.Error
+		if ok := errors.As(err, &pErr); ok {
+			switch pErr.Code {
+			case "app_not_found":
+				cmdutil.Fatalf("app not found: %s", appSlug)
+			case "validation":
+				var details platform.ValidationDetails
+				err := json.Unmarshal(pErr.Detail, &details)
+				if err != nil {
+					cmdutil.Fatalf("failed to deploy: %v", err)
+				}
+				switch details.Field {
+				case "commit":
+					cmdutil.Fatalf("could not find commit: %s. Is it pushed to the remote repository?", commit)
+				case "branch":
+					cmdutil.Fatalf("could not find branch: %s. Is it pushed to the remote repository?", branch)
+				case "env":
+					cmdutil.Fatalf("could not find environment: %s/%s", appSlug, envName)
+				}
+			}
+		}
+		if err != nil {
+			cmdutil.Fatalf("failed to deploy: %v", err)
+		}
+		url := fmt.Sprintf("https://app.encore.dev/%s/deploys/%s/%s", appSlug, rollout.EnvName, strings.TrimPrefix(rollout.ID, "roll_"))
+		switch format {
+		case "text":
+			fmt.Println(aurora.Sprintf("\n%s %s\n", aurora.Bold("Started Deploy:"), url))
+		case "json":
+			output, _ := json.Marshal(map[string]string{
+				"id":  strings.TrimPrefix(rollout.ID, "roll_"),
+				"env": rollout.EnvName,
+				"app": appSlug,
+				"url": url,
+			})
+			fmt.Println(string(output))
+		}
+	},
+}
+
+func init() {
+	alphaCmd.AddCommand(deployAppCmd)
+	deployAppCmd.Flags().StringVar(&appSlug, "app", "", "app slug to deploy to (default current app)")
+	deployAppCmd.Flags().StringVarP(&envName, "env", "e", "", "environment to deploy to (default primary env)")
+	deployAppCmd.Flags().StringVarP(&commit, "commit", "c", "", "commit to deploy")
+	deployAppCmd.Flags().StringVarP(&branch, "branch", "b", "", "branch to deploy")
+	deployAppCmd.Flags().StringVarP(&format, "format", "f", "text", "format of output. One of (text, json)")
```

Comment:
We have `cmdutil.Oneof` for this use case which does some niceties like autocomplete and validation